llvm/lib/LTO/ThinLTOCodeGenerator.cpp

```diff
@@ -1557,6 +1761,273 @@ void ThinLTOCodeGenerator::run() {
+      });
+    };
+
+    std::mutex WriteQueueMutex;
+    std::queue<int> WriteQueue;
+    auto EnqueueCacheWrite = [&](int count) {
+      std::scoped_lock Lock(WriteQueueMutex);
+      WriteQueue.push(count);
+      if (HandledCacheReads.check(AllModules)) {
+        while (!WriteQueue.empty()) {
+          WriteCache(WriteQueue.front());
+          WriteQueue.pop();
+        }
+      }
+    };
+
+    // Query the cache.
+    for (auto IndexCount : ModulesOrdering) {
+      auto &Mod = Modules[IndexCount];
+      auto ModuleIdentifier = Mod->getName();
+
+      std::string OutputPath = computeThinLTOOutputPath(
+          IndexCount, SavedObjectsDirectoryPath, TMBuilder);
+
+      auto &CacheEntry = Infos[IndexCount].Entry = createAsyncModuleCacheEntry(
+          GetCancelTok, PutCancelTok, CacheOptions, *Index, ModuleIdentifier,
+          OutputPath, ImportLists[ModuleIdentifier],
+          ExportLists[ModuleIdentifier], ResolvedODR[ModuleIdentifier],
+          ModuleToDefinedGVSummaries[ModuleIdentifier], OptLevel, Freestanding,
+          TMBuilder,
+          [&CacheLogOS](llvm::function_ref<void(raw_ostream & OS)> Log) {
+            if (CacheLogging)
+              CacheLogOS.applyLocked(Log);
+            else {
+              raw_null_ostream OS;
+              Log(OS);
+            }
+          });
+
+      CacheEntry->tryLoadingBuffer([&, IndexCount,
+                                    GetCancelTok](auto ErrOrBuffer) {
+        auto CancelTokHandle = GetCancelTok->acquireHandle();
+        if (!CancelTokHandle)
+          return;
+
+        bool CacheHit{ErrOrBuffer};
+        unsigned CacheBits = MS_CacheLookupDone;
+
+        if (CacheHit) {
+          Infos[IndexCount].CachedBuffer = std::move(ErrOrBuffer.get());
+          CacheBits |= MS_CacheLookupHit;
+        }
+
+        unsigned ComputationState = Infos[IndexCount].State.fetch_or(CacheBits);
+
+        // Consider writing the object file.
+        if (DeterministicCheck) {
+          // With deterministic checks, writing the object file is always
+          // responsibility of the compute task.
+        } else {
+          // Without deterministic checks, writing the object file is
+          // responsibility of whoever comes up with usable result first.
+          if (!(ComputationState & MS_ComputationDone) && CacheHit)
+            WriteObject(IndexCount, /*CacheHit=*/true);
+        }
+
+        HandledCacheReads.apply(Increment);
+
+        // Enqueueing cache write is responsibility of whoever lost the race.
+        if (DeterministicCheck) {
+          if (ComputationState & MS_ComputationDone)
+            EnqueueCacheWrite(IndexCount);
+        } else {
+          if (ComputationState & MS_ComputationDone) {
+            if (!CacheHit)
+              EnqueueCacheWrite(IndexCount);
+            else
+              EnqueueCacheWrite(-1);
+          }
+        }
+      });
+    }
+
+    // Parallel optimizer + codegen
```

Comment:
Is this block same as blow?

Reply:
No, this block is pretty different from the existing logic. Only the chunk of code between line 1960 and 1986 is identical, and I think it's still better to duplicate that in order to make the downstream diff more self-contained.